### PR TITLE
feat(site): Ask for version name and if it is active when publishing a new version on editor

### LINKED ISSUE #6124: template editor: custom version name & promote to default
I'd love to be able to make non-default template versions as well as specify a custom name. With #5865/#3775, we could also let the user specify a message.

Screenshot 2023-02-08 at 3 44 44 PM

### LINKED ISSUE #6169: template editor: clicking `Publish New Version` button in template edit screen doesn't take user back to template screen
clicking the `Publish New Version` button doesn't take the user away from the screen. i have to manually navigate out of the edit view and click back into the template to ensure it's been updated. the button does, however, publish the new template version in the back-end.

Screenshot 2023-02-11 at 10 18 22 PM

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
Demo:

https://user-images.githubusercontent.com/3165839/227322745-24b91562-2623-4514-bbde-25ab7f97bae1.mov

Fix #6169 
Close #6124

### CODE REVIEW

File: site/src/api/api.ts [hunk 1]
```diff
@@ -1020,3 +1031,26 @@ const getMissingParameters = (
...
+    const proto = location.protocol === "https:" ? "wss:" : "ws:"
+    const socket = new WebSocket(
+      `${proto}//${location.host}/api/v2/templateversions/${versionId}/logs?follow=true`,
+    )
+    socket.binaryType = "blob"
+    socket.addEventListener("message", (event) =>
+      onMessage(JSON.parse(event.data) as TypesGen.ProvisionerJobLog),
+    )
+    socket.addEventListener("error", () => {
+      reject(new Error("Connection for logs failed."))
```
Comment (Kira-Pilot) [inline site/src/api/api.ts:1049]: Will the socket close automatically after a couple of failures?

Comment (BrunoQuaresma) [reply site/src/api/api.ts:1049]: Good question, let me investigate that.

File: site/src/components/TemplateVersionEditor/PublishTemplateVersionDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        await form.submitForm()
+      }}
+      hideCancel={false}
+      type="success"
+      cancelText="Cancel"
+      confirmText="Publish"
+      title="Publish new version"
+      description={
+        <Stack>
+          <p>You are about to publish a new version of this template.</p>
```
[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply site/src/components/TemplateVersionEditor/PublishTemplateVersionDialog.tsx:56]: I always forget translations. Sorry 🙏

File: site/src/components/TemplateVersionEditor/TemplateVersionEditor.tsx [hunk 3]
```diff
@@ -156,233 +166,247 @@ export const TemplateVersionEditor: FC<TemplateVersionEditorProps> = ({
...
+                open={createFileOpen}
+                onClose={() => {
+                  setCreateFileOpen(false)
+                }}
+                checkExists={(path) => existsFile(path, fileTree)}
+                onConfirm={(path) => {
+                  setFileTree((fileTree) => createFile(path, fileTree, ""))
+                  setActivePath(path)
+                  setCreateFileOpen(false)
+                  setDirty(true)
```
Comment (Kira-Pilot) [inline site/src/components/TemplateVersionEditor/TemplateVersionEditor.tsx:250]: These can fire asynchronously - is that okay?

File: site/src/components/TemplateVersionEditor/PublishTemplateVersionDialog.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kira-Pilot) [reply site/src/components/TemplateVersionEditor/PublishTemplateVersionDialog.tsx:56]: no worries - i often forget too

**Review by Kira-Pilot — APPROVED** [review]

File: site/src/api/api.ts (hunk 1 above)
Comment (Kira-Pilot) [reply site/src/api/api.ts:1049]: We could maybe make use of a `useEffect` in the component itself to make sure the socket is cleaned up when the user navigates away from the page, even if the logs haven't finished streaming.

Comment (BrunoQuaresma) [reply site/src/api/api.ts:1049]: if this is used on react-query or xstate - they should cleanup those things but I will check. This was already working, I just moved into a function on API to make it easier to mock it.

**Comment by BrunoQuaresma** [comment]
> This is a really cool feature - great job! Do we need any validation on the version name? I imagine we might get a conflict if we have two of the same name.

It is a good question. Let me try and see what happens.

File: site/src/api/api.ts (hunk 1 above)
Comment (BrunoQuaresma) [reply site/src/api/api.ts:1049]: It is not closing the connection!!! This is a thing that I didn't expect.

File: site/src/components/TemplateVersionEditor/PublishTemplateVersionDialog.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply site/src/components/TemplateVersionEditor/PublishTemplateVersionDialog.tsx:56]: > This is a really cool feature - great job! Do we need any validation on the version name? I imagine we might get a conflict if we have two of the same name.

You found a bug on BE haha. Will take a look on that.

File: site/src/api/api.ts (hunk 1 above)
Comment (BrunoQuaresma) [reply site/src/api/api.ts:1049]: Related to the websocket, I think this is happening for all the other ws connections that we have on the site. I created a ticket to take it separately  https://github.com/coder/coder/issues/6782

Comment (Kira-Pilot) [reply site/src/api/api.ts:1049]: Thanks for doing that!

**Pull request merged**